x/posts/keeper/posts.go

```diff
@@ -0,0 +1,183 @@
+
+// DeleteNextPostID removes the post id key for the given subspace
+func (k Keeper) DeleteNextPostID(ctx sdk.Context, subspaceID uint64) {
+	store := ctx.KVStore(k.storeKey)
+	store.Delete(types.NextPostIDStoreKey(subspaceID))
+}
+
+// --------------------------------------------------------------------------------------------------------------------
+
+// ValidatePostReference checks the post reference to make sure that the referenced
+// post's author has not blocked the user referencing the post
+func (k Keeper) ValidatePostReference(ctx sdk.Context, postAuthor string, subspaceID uint64, referenceID uint64) error {
+	// Make sure the referenced post exists
+	referencedPost, found := k.GetPost(ctx, subspaceID, referenceID)
+	if !found {
+		return sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "post with id %d does not exist", referenceID)
+	}
+
+	// Make sure the original author has not blocked the post author
+	if k.HasUserBlocked(ctx, referencedPost.Author, postAuthor, subspaceID) {
+		return sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "author of post %d has blocked you", referenceID)
+	}
+
+	return nil
+}
+
+// ValidatePostReply checks the original post reply settings to make sure that
+// only specified users can answer to the post
+func (k Keeper) ValidatePostReply(ctx sdk.Context, postAuthor string, subspaceID uint64, referenceID uint64) error {
+	replyPost, found := k.GetPost(ctx, subspaceID, referenceID)
+	if !found {
+		return sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "post with id %d does not exist", referenceID)
+	}
+
+	switch replyPost.ReplySettings {
+	case types.REPLY_SETTING_FOLLOWERS:
+		// We need to make sure that a relationship between post author -> original author exists
+		if !k.HasRelationship(ctx, postAuthor, replyPost.Author, subspaceID) {
+			return sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "only followers of the author can reply to this post")
+		}
+
+	case types.REPLY_SETTING_MUTUAL:
+		// We need to make sure that both relationships exist (post author -> original author and original author -> post author)
+		if !k.HasRelationship(ctx, postAuthor, replyPost.Author, subspaceID) || !k.HasRelationship(ctx, replyPost.Author, postAuthor, subspaceID) {
+			return sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "only mutual connections of the author can reply to this post")
+		}
+
+	case types.REPLY_SETTING_MENTIONS:
+		// We need to check each mention of the original post
+		if !replyPost.IsUserMentioned(postAuthor) {
+			return sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "only mentioned users can reply to this post")
+		}
+	}
+
+	return nil
+}
+
+// ValidatePost validates the given post based on the current params, returning an error if anything is wrong
+func (k Keeper) ValidatePost(ctx sdk.Context, post types.Post) error {
+	params := k.GetParams(ctx)
+
+	// Validate the conversation reference
+	if post.ConversationID != 0 {
+		err := k.ValidatePostReference(ctx, post.Author, post.SubspaceID, post.ConversationID)
```

Comment:
The conversation post has the same function with referenced post. Should we merge it into referenced post with `conversation` type?

Reply:
The idea of having a conversation is to allow replicating something similar to Twitter conversations. The `ConversationID` represents the first post of the conversation (which can continue as a series of posts). A `PostReference`, on the other hand, should be used for reposts, quotes and other things. But the conversation id should remain separate to easily identify all the posts of the same conversation in my opinion